src/wkt/tests/message_with_enum.rs

```diff
@@ -0,0 +1,116 @@
+    fn test_singular(input: Value, want: TestEnum) -> Result {
+        let got = serde_json::from_value::<MessageWithEnum>(input)?;
+        assert_eq!(got.singular, want, "{got:?}");
+        let roundtrip = serde_json::from_value(serde_json::to_value(&got)?)?;
+        assert_eq!(got, roundtrip);
+        Ok(())
+    }
+
+    #[test_case(json!({}))]
+    #[test_case(json!({"singular": null}))]
+    #[test_case(json!({"singular": 0}))]
+    #[test_case(json!({"singular": "TEST_ENUM_UNSPECIFIED"}))]
+    fn test_singular_default(input: Value) -> Result {
+        let got = serde_json::from_value::<MessageWithEnum>(input)?;
+        assert_eq!(got, MessageWithEnum::default());
+        let output = serde_json::to_value(&got)?;
+        assert_eq!(output, json!({}));
+        Ok(())
+    }
+
+    #[test_case(json!({"optional": "RED"}), TestEnum::Red)]
+    #[test_case(json!({"optional": 1}), TestEnum::Red)]
+    fn test_optional(input: Value, want: TestEnum) -> Result {
+        let got = serde_json::from_value::<MessageWithEnum>(input)?;
+        assert_eq!(got.optional, Some(want), "{got:?}");
+        let roundtrip = serde_json::from_value(serde_json::to_value(&got)?)?;
+        assert_eq!(got, roundtrip);
+        Ok(())
+    }
+
+    #[test_case(json!({}))]
+    #[test_case(json!({"optional": null}))]
+    fn test_optional_default(input: Value) -> Result {
+        let got = serde_json::from_value::<MessageWithEnum>(input)?;
+        assert_eq!(got, MessageWithEnum::default());
+        let output = serde_json::to_value(&got)?;
+        assert_eq!(output, json!({}));
+        Ok(())
+    }
+
+    #[test]
+    fn test_repeated() -> Result {
+        let msg = MessageWithEnum::new().set_repeated([TestEnum::Red, TestEnum::Unspecified]);
+        let got = serde_json::to_value(&msg)?;
+        let want = json!({"repeated": [1, 0]});
+        assert_eq!(want, got);
+        let roundtrip = serde_json::from_value(serde_json::to_value(&msg)?)?;
+        assert_eq!(msg, roundtrip);
+        Ok(())
+    }
+
+    #[test_case(json!({}))]
+    #[test_case(json!({"repeated": []}))]
+    #[test_case(json!({"repeated": null}))]
+    fn test_repeated_default(input: Value) -> Result {
+        let got = serde_json::from_value::<MessageWithEnum>(input)?;
+        assert_eq!(got, MessageWithEnum::default());
+        let output = serde_json::to_value(&got)?;
+        assert_eq!(output, json!({}));
+        Ok(())
+    }
+
+    #[test]
+    fn test_map() -> Result {
+        let input = json!({"map": { "favorite": "RED", "not-so-much": "GREEN"}});
+        let got = serde_json::from_value::<MessageWithEnum>(input.clone())?;
+        let want = MessageWithEnum::new().set_map([
+            ("favorite", TestEnum::Red),
+            ("not-so-much", TestEnum::Green),
+        ]);
+        assert_eq!(want, got);
+        let roundtrip = serde_json::from_value(serde_json::to_value(&got)?)?;
+        assert_eq!(got, roundtrip);
```

Comment:
nit: Just do `to_value(&got)` and compare against `input`? I assume that is why we cloned `input` above.

Reply:
Cannot do that here.  to `to_value(&got) == json!({"map: { "favorite": 1, "not-so-much": 2 })`.

I will  remove the `clone()`.

The tests are written as they are because in general JSON -> `type` -> JSON does not get you the same thing (that is true from most primitive types too), while `type` -> JSON -> `type` does give you the same thing.

Reply:
Done.